### ISSUE TITLE
Completely include (or exclude) features sitting on the edge of a candidate cluster

### ISSUE THREAD

**Issue by jorgecnavarrom** [antismash/antismash#684]
**Is your feature request related to a problem? Please describe.**
It seems like features that are on the border of a protocluster/candidate cluster are not included in the antiSMASH region results 

**Describe the solution you'd like**
Resize the protocluster so that it fully includes features on the border (default) or fully excludes them

**Describe alternatives you've considered**
Include partial features (but less ideal)

**Additional context**
For example, an antiSMASH [run on CP001042](https://antismash.secondarymetabolites.org/upload/bacteria-512aee5e-7a51-40cc-8e37-ac6602436ffd/index.html#r1c1) gives a single RiPP-like region. The region overview shows 6 CDSs (Npun_ER021 - Npun_ER026), with about 2kbp of empty space at the 5' end. However, in the GenBank summary file, there is another feature (Npun_EF019) in the border of the region, not shown at all.
```
     gene            16826..19885
                     /locus_tag="Npun_EF019"
     CDS             16826..19885
                     /codon_start=1
                     /db_xref="InterPro:IPR014015"
                     /locus_tag="Npun_EF019"
                     /note="PFAM: Helicase superfamily 3; KEGG: ana:all9003
                     unknown protein"
                     /product="Helicase superfamily 3"
                     /protein_id="ACC85487.1"
                     /transl_table=11
                     /translation="MVAILKGQETVGGSAMPGAFASFDIRNFQDQLEPSKEKGKYICPV
                     ...
                     INYDSKNGIVTVATSQGNRDASLCETFVID"
     protocluster    18867..26419
```

**Comment by SJShaw** [comment]
Unfortunately, this is problematic whichever method is chosen. Gene overlaps of even one base, whether it's due to bad gene calling, bad annotations, frame shifts, sequencing/assembly problems, etc. There's quite a few examples of overlapping genes in RefSeq entries, e.g. in NC_003888, there's quite a lot of these overlaps, with this particular example pair having length of 3.5k nt in just two genes: 
![image](https://github.com/antismash/antismash/assets/1700735/ec93bbb5-fb39-4717-8292-559755852df8)

Another example, though the lengths of each gene in this one are much smaller, it's much more obvious due to the strands being reversed:
![image](https://github.com/antismash/antismash/assets/1700735/629e1413-6122-4234-bf1a-49cf5ece0e50)

**Exclusion**
If a non-core gene, i.e. one in the neighbourhood, overlaps with a core gene, it will already be partially present in the protocluster core feature, because it can't be included in the core and can't be fully excluded because then the core gene would be excluded. If we did, it would result in the neighbourhood smaller than the core.

While it's unlikely that a single gene beside a core gene would be long enough for a neighbourhood to intersect it, this overlap problem is transitive. If core gene 1 overlaps with non-core gene 2, which overlaps with 3, ... which overlaps with N, then _all_ those genes must be excluded and then we're back to not being able to exclude the neighbouring gene due to the core overlap.

**Inclusion**
Similarly, the inclusion case is transitive. If a gene would be cut by the neighbourhood and overlaps with a gene currently excluded, when the neighbourhood is extended to cover the first, the second will be cut. The neighbourhood would then have to continually extend until an intergenic region is hit. This would result in some cases where rules with small neighbourhoods (e.g. `RiPP-like`, with a neighbourhood of 5k nt) could easily double or triple, which in turn would look very strange when the genes themselves don't look like overlaps.

Even worse are the genes which are 20-30kb by themselves, which is as long or longer than the neighbourhood range in almost every detection rule we have. Extending the neighbourhood for one of these becomes problematic very, very quickly.


This is why we went with the very arbitrary and constant length (barring contig edges, of course). Fungi are easier this way, of course, but in bacteria there's just.

**Comment by jorgecnavarrom** [comment]
The overlapping of genes is an interesting problem indeed. The simplest solution I can think of is to extend the region only once, until it fully includes any gene that intersects the original _region_ border, but don't apply the same logic again if the included gene's borders intersects any other gene. The only extreme case then would be the original border falling exactly on a gene overlap, and having two include both genes.

**Comment by SJShaw** [comment]
The region borders are defined by the max/min of overlapping protocluster (and/or subregion, if present) borders, so we can't know what they are until the inclusions/exclusions are made to protoclusters.

If we extend region borders after the protoclusters are all defined, the protoclusters will still cut genes, but we'll also introduce the possible region overlaps, which would be even more complicated to handle.

It's a very tricky problem.

**Comment by jorgecnavarrom** [comment]
you're right, sorry! I wasn't very careful with my comment above when I referred to "regions". 

What I mean is: (if I remember correctly) after having a hit with a rule, the protocluster feature is created and its span extends according to the `EXTEND` parameter in the rule. At this point, antiSMASH could keep extending if the neighborhood's border is in the middle of a gene until the end of the gene. Leaving out possible genes that (chain-)overlap with the border gene would still be problematic, but at least the border gene would correctly show up now in the results (both the figure and the genbank). Then hopefully the downstream steps would not be affected (e.g. candidate cluster definition, etc.)

**Comment by SJShaw** [comment]
`EXTEND` was the antiSMASH 4 and before term, `EXTENDERS` in current antiSMASH rules are profiles that will extend the core, once the core is found.

Your version still has the problem of having a neighbourhood extended to much longer than it normally is, even with a single gene, which I mentioned in the Inclusion section in my first reply. As an example, a 20-30kb modular (e.g. a modular PKS/NRPS gene, which almost certainly won't be relevant to a RiPP, but could make the neighbourhood 6 times longer just with that one gene.

**Comment by SJShaw** [comment]
It occurred to me that we might be talking past each other a bit. If your goal is to have the visualisation *show* that partial gene, rather than leaving it empty, that we can do without changing the protocluster borders. We'll just have to come up with a good indicator that they're truncated.

**Comment by jorgecnavarrom** [comment]
> Your version still has the problem of having a neighbourhood extended to much longer than it normally is, even with a single gene, which I mentioned in the Inclusion section in my first reply. As an example, a 20-30kb modular (e.g. a modular PKS/NRPS gene, which almost certainly won't be relevant to a RiPP, but could make the neighbourhood 6 times longer just with that one gene.

Yes, this could be the case in some situations. I wouldn't say that's necessarily bad, as actual borders may look very different in reality anyway (shorter or longer). 

> It occurred to me that we might be talking past each other a bit. If your goal is to have the visualisation show that partial gene, rather than leaving it empty, that we can do without changing the protocluster borders. We'll just have to come up with a good indicator that they're truncated.

This would also be a good solution as the region shown by antiSMASH would more accurately describe the data behind it. I guess the (region) genbank file could also have truncated features, but it could be retrieved from the whole-genome gbk if needed.

**Comment by SJShaw** [comment]
Yeah, it's the visual inconsistency in neighbourhood distances that's going to be really awkward.

Let's go with adding the truncated genes to the HTML visualisation for starters and see if it works well enough.

**Comment by jorgecnavarrom** [comment]
Great!

**Comment by zreitz** [comment]
I'm in favor of adding partial CDSs to the HTML, as I don't like that the current method can imply some pretty large intergenic distances that aren't real, ex: [this old aSDB record that bugged me](https://antismash-dbv3.secondarymetabolites.org/output/GCF_000022685.1/#r1c3).

A complication, but hopefully not too much of one, will be that if the partial CDSs are in the HTML, I think they should also be in the region GBK file. And if they're in the GBK, antiSMASH should still be able to accept its output as input, whereas currently you receive `ERROR: feature translation extends out of record`.

**Comment by SJShaw** [comment]
> A complication, but hopefully not too much of one, will be that if the partial CDSs are in the HTML, I think they should also be in the region GBK file. And if they're in the GBK, antiSMASH should still be able to accept its output as input, whereas currently you receive `ERROR: feature translation extends out of record`. Accepting partial features would generally be nice when running aS on incomplete records, IMO.

I was just coming back to this issue as part of some other javascript changes I was making, since I could bundle them together, and thought I'd respond to this part since it's not quite related to the visualisation.

When it comes to confusing or invalid files (e.g. translations that are much too long for the feature or long record identifiers merging with locations), the general principle is to abort and let the user check and correct, if relevant, their input, as it's much, much worse for us to show misleading results than to fail outright.

There is existing support for inputs to antiSMASH that contain partial genes/CDS features, provided they're annotated correctly the relevant ambiguous coordinate syntax in the GenBank file format (https://www.ncbi.nlm.nih.gov/genbank/samplerecord/#BaseSpanA). 

It will be quite easy to truncate these partial features in the output files using the same syntax, which will then be able to come right back in to antiSMASH.

**Comment by zreitz** [comment]
> There is existing support for inputs to antiSMASH that contain partial genes/CDS features, provided they're annotated correctly the relevant ambiguous coordinate syntax in the GenBank file format (https://www.ncbi.nlm.nih.gov/genbank/samplerecord/#BaseSpanA).

Really? Downloaded selections from the [NCBI genome viewer](https://www.ncbi.nlm.nih.gov/projects/sviewer/?Db=gene&DbFrom=protein&Cmd=Link&noslider=1&id=NC_012808.1&from=4209912&to=4232908) fail with partial CDS, despite having ambiguous locations.

(Renamed to .txt to enable uploading here)
[NC_012808.1[4213998..4228882].txt](https://github.com/antismash/antismash/files/14196951/NC_012808.1.4213998.4228882.txt)
`ERROR: feature translation extends out of record: MEXAM1_RS19435`
[NC_012808.1[4218240..4228833].txt](https://github.com/antismash/antismash/files/14197001/NC_012808.1.4218240.4228833.txt)
`ERROR: translation longer than location allows: 1782 > 1689: MEXAM1_RS19450`

**Comment by SJShaw** [comment]
> > There is existing support for inputs to antiSMASH that contain partial genes/CDS features, provided they're annotated correctly the relevant ambiguous coordinate syntax in the GenBank file format (https://www.ncbi.nlm.nih.gov/genbank/samplerecord/#BaseSpanA).
> 
> Really?

[Yes](https://github.com/antismash/antismash/blob/master/antismash/common/secmet/features/cds_feature.py#L49-L57), at least for features with a well defined start coordinate.

Allowing ambiguous start locations becomes very awkward very quickly. Even if we allow them in general, we'll likely still have to reject any non-trivial cases that don't have a translation defined in the input.

> Downloaded selections from the [NCBI genome viewer](https://www.ncbi.nlm.nih.gov/projects/sviewer/?Db=gene&DbFrom=protein&Cmd=Link&noslider=1&id=NC_012808.1&from=4209912&to=4232908) fail with partial CDS, despite having ambiguous locations.
> 
> (Renamed to .txt to enable uploading here) [NC_012808.1[4213998..4228882].txt](https://github.com/antismash/antismash/files/14196951/NC_012808.1.4213998.4228882.txt) `ERROR: feature translation extends out of record: MEXAM1_RS19435` [NC_012808.1[4218240..4228833].txt](https://github.com/antismash/antismash/files/14197001/NC_012808.1.4218240.4228833.txt) `ERROR: translation longer than location allows: 1782 > 1689: MEXAM1_RS19450`

That's true, these full translations of genes with a partial location, cut by a contig edge, are still blocked in inputs, because we can't annotate any results outside the record in a genbank file (arguably we could, but a location of [<0:<0], for example, is very far from ideal). Partially analysing them would be misleading, as would partially annotating them if the full translation was used. In our outputs with partial genes, we could very easily truncate those relevant translations and they'd be able to be fed back in as a separate input, however that will still lead to some issues with things like NCBI locus tags in our output having different translations from those in the NCBI entries.

For genbank outputs, it won't be clean, whichever way it goes. Drawing/visualisation, on the other hand, should be quite manageable.
